---
title: "Practicum 2 Analysis"
author:
- Peter Claussen
- Maggie Germundson
- Stephanie Liebl
date: "11/2/2021"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(cluster)
library(ggdendro)
set.seed(2)
```

# Introduction

We have been give a large data file, consisting of expression levels of RNA for different genes. The mRNA were obtained from multiple cell lines. The client requests that we use RNA expression patterns to group cell lines. These data are expected to be very sparse. 

### Notation

A brief note on notation. Cell lines are given an arbitrary sequence of nucleotides, i.e. `AAACATACAACCAC.1` while gene names follow standard human gene nomenclature. In this data set, there are 2700 cell lines and 13714 genes.
The data are provided as a table in Excel format. We assume value represent relative expression levels. 

```{r,echo=FALSE,include=FALSE}
if(!file.exists('single.cell.Rda')) {
  cell.dat <- read.csv('single_cell.csv',header=TRUE)
  save(cell.dat,file='single.cell.Rda')
} else {
  load(file='single.cell.Rda')
}
#first column are the gene identifiers
genes <- cell.dat[,1]
#first row are barcodes for cell lines
bar.code <- cell.dat[1,]
cell.dat <- cell.dat[,-1]
dim(cell.dat)
```

Our first task will be to screen the number of candidate genes to a manageable level. We start this process by considering the incidence of gene expression. We determine where a gene is expressed by any given cell line, regardless of the level of expression. From this, we identify 4 potential groups of genes.

```{r,echo=FALSE}
incidence.dat <- cell.dat
incidence.dat[incidence.dat>0.01] <- 1

incidences <- apply(incidence.dat,1,mean)
hist(incidences, main='Incidence of gene expression',xlab='Proportion of cell lines')

incidence.dat <- incidence.dat[incidences>0.1 & incidences<0.2,]
genes.incidence <- genes[incidences>0.1 & incidences<0.2]
dim(incidence.dat)
```

First, we consider the group where the median expression level is greater than 0. This identifies those genes that are expressed in at least half of the cell lines, and we call these **commonly expressed genes**. From these, we select those genes whose expression patterns are approximately normally distributed (or, at least, symmetrically distributed), define as those genes whose mean gene expression levels are approximately equal to the median gene expression level. These are the **most commonly expressed** genes, and we will use these genes to illustrate the analysis method.

We wish to avoid using genes that are rarely expressed; for this analysis, we will define **rarely expressed** genes as those that are expressed in less than 10% of cell lines. We also define **uncommon genes** as those that are expressed in between 10-20% of cell lines. 


```{r,eval=FALSE}
row.max <- apply(cell.dat,1,max)
hist(row.max)
```


```{r,echo=FALSE,eval=FALSE}
distances <- dist(t(cell.dat), method="euclidean")
clusters <- hclust(distances,method="ward.D")
ggdendrogram(clusters,rotate=TRUE)
```

```{r,eval=FALSE,fig.width=6,fig.height=8}
random.samples <- sample(1:dim(cell.dat)[1],12)
random.samples
par(mfrow=c(4,3))
for(i in random.samples) {
  tmp <- unlist(cell.dat[i,])
  hist(tmp,main=paste("Histogram,",genes[i]))
}
```

\pagebreak
## Commonly expressed genes

We find 273 genes that can be classified as commonly expressed.

```{r,echo=FALSE}
row.median <- apply(cell.dat,1,median)
dup <- row.median
summary(row.median)
dup <- row.median[row.median>0]
#hist(dup)
work.dat <- cell.dat[row.median>0,]
genes.common <- genes[row.median>0]
dim(work.dat)
```

We plot, for illustrative purposes, the histograms of expression levels of a random sample of 12 commonly expressed genes. We see that many of the gene expression levels are approximately normally distributed, but many have an excess of zeros. This may complicate further analysis, so we screen for approximate normality as described above.

```{r,fig.width=6,fig.height=8}
print(common.sample <- sample(genes.common,12))
common.mask <- genes %in% common.sample
common.dat <- cell.dat[common.mask,]
common.names <- genes[common.mask]
par(mfrow=c(4,3))
for(i in 1:dim(common.dat)[1]) {
  tmp <- unlist(common.dat[i,])
  hist(tmp,main=paste("Histogram, ",common.names[i]))
}
```

```{r,eval=FALSE,fig.width=6,fig.height=8}
distances <- dist(t(work.dat), method="euclidean")
clusters <- hclust(distances,method="ward.D")
ggdendrogram(clusters,rotate=TRUE)
```

```{r,eval=FALSE,fig.width=6,fig.height=8}
heatmap(t(work.dat))
```

\pagebreak
## Most commonly expressed genes

We compute both mean and median expression levels for the 273 commonly expressed genes, and select those where the ratio of mean/median is between 0.95 and 1.05 (in other words, the mean is about plus/minus 5% of the median.)

```{r,echo=FALSE}
row.means <- apply(cell.dat,1,mean)
row.median <- apply(cell.dat,1,median)
```

```{r}
ratio <- row.means/row.median
mask <- ratio<1.05 & ratio>0.95
sum(mask)
work.dat <- cell.dat[mask,]
genes.work <- genes[mask]
dim(work.dat)
```

Again, we plot expression levels of a random sample the most commonly expressed genes. Note these expression levels exclude most cases of zero-inflated normal distributions.

```{r,echo=FALSE,fig.width=6,fig.height=8}
print(work.sample <- sample(genes.work,12))

work.mask <- genes %in% work.sample
common.dat <- cell.dat[work.mask,]
common.names <- genes[work.mask]
par(mfrow=c(4,3))
for(i in 1:dim(common.dat)[1]) {
  tmp <- unlist(common.dat[i,])
  hist(tmp,main=paste("Histogram, ",common.names[i]))
}
```

# Gene clusters

We create a dendrogram and a heat map illustrating cell groups based on expression levels of the most commonly expressed genes. Our next step will be deciding how to use the tree to assign cell lines to related groups.

```{r}
distances <- dist(t(work.dat), method="euclidean")
clusters <- hclust(distances,method="ward.D")
```

```{r,eval=FALSE,fig.width=6,fig.height=8}
ggdendrogram(clusters,rotate=TRUE)
```


```{r,eval=FALSE,fig.width=6,fig.height=8}
heatmap(t(work.dat))
```

To determine a reasonable number of cell line groups, we use the elbow method. Briefly, we compute, for each choice of the number of groups $k$, a within-group sums of squares (WSS). This number will generally decrease with the number of groups, but the rate of change of WSS declines with group count, and increasing group number produces less improvement in WSS. We look for an 'elbow' in the plot of WSS against group number to find the optimal number of groups.

```{r,echo=FALSE}
#Code borrowed from Practical Data Science with R 
sqr_edist <- function(x,y) {
  sum((x-y)^2)
}
wss_cluster <- function(clustermat) {
  c0 <- colMeans(clustermat)
  sum(apply(clustermat, 1, FUN= function(row){sqr_edist(row,c0)}))
}
wss_total <- function(dmatrix,labels) {
  wsstot <- 0
  k <- length(unique(labels)) 
  for(i in 1 :k) {
    wsstot <- wsstot + wss_cluster(subset(dmatrix, labels==i))
  }
  wsstot
}

get_wss <- function(dmatrix, max_clusters) {
  wss = numeric(max_clusters)
  wss[1] <- wss_cluster(dmatrix)
  d <- dist(dmatrix, method='euclidean')
  pfit <- hclust(d, method='ward.D')
  for(k in 2:max_clusters) {
    labels <- cutree(pfit,k=k)
    wss[k] <- wss_total(dmatrix, labels)
  }
  wss
}
```

```{r,echo=FALSE}
library(ggplot2)
kmax=20
cluster_meas <- data.frame(nclusters=1:kmax,
                             wss = get_wss(work.dat,kmax))
breaks <- 1:kmax
ggplot(cluster_meas,aes(x=nclusters,y=wss)) + geom_point() + geom_line() +
  scale_x_continuous(breaks=breaks)
```

In the above graph, we have two candidate groups. The most obvious 'elbow' is at $k=3$, but there appears an additional break at $k=6$. For this analysis, we choose 6 groups; we expect that there are more than three distinct cell lines in these data.

We assign groups to data but 'cutting' the dendrogram at the point where there are 6 distinct branches, then assign a group number to each cell line. The assignment of cell groups to group lines is provided in the file `cellgroups.csv`

```{r,echo=FALSE}
grouped.work.dat <- data.frame(t(work.dat))
names(grouped.work.dat) <- genes.work
grouped.work.dat$Group <- cutree(clusters,k=6)
grouped.work.dat$Group <- as.factor(grouped.work.dat$Group)
```

## Comparison of gene expression levels by group.

We have identified 81 genes that are most commonly expressed in these cell lines, and have created groups that are determined by expression patterns of these genes. We now analysis each gene, independently, to determine if there are differences in express for each group of cell lines. We do this by simple one-way analysis of variance, and report only those cell lines where the between group F-ratio is significant. Since there will be 81 tests of significance, we apply an FDR correction to the p-values found for each AOV. A table listing gene names along with unadjusted and adjusted p-values is provided as `genesofinterest.csv`. This table contains only those genes with adjusted between groups p-value less than 0.05.

```{r}
#PMC Stephanie will write a loop to create a table of gene names and whether they are differentially expressed
library(stringr)
library(data.table)

colnames.work.dat <-colnames(grouped.work.dat)
temp <- str_replace_all(colnames.work.dat, '-', '_')
temp.work.dat <- grouped.work.dat
colnames(temp.work.dat) <- temp

tbls <- data.frame(x=1:81)

for(i in 1:(ncol(temp.work.dat)-1)){
  fmla <- as.formula(paste(temp[i], "~ Group"))
  aov.tmp <-aov(fmla,data=temp.work.dat)
  aov.tbl <- summary(aov.tmp)
  tbls$pvalue[i] <- aov.tbl[[1]][1,'Pr(>F)']
  tbls$fmla[i] <- temp[i]
}
# subset to retrieve only the p-values less than 0.05^n = 0.05^81 = very small
#sig.pvals <- tbls[which(tbls$pvalue<=0.05^81),]

#PMC - use R to give use adjusted p-values, using Benjamin-Hochberg FDR
#tbls$p.adjust <- p.adjust(tbls$pvalue, method = "fdr")
tbls$p.adjust <- p.adjust(tbls$pvalue, method = "bonferroni")
sig.pvals <- tbls[which(tbls$p.adjust<=0.05),]
fwrite(sig.pvals, file='genesofinterest.csv')

# write cell/group to excel
cells <- rownames(temp.work.dat)
group <- temp.work.dat$Group
cellgroups <- as.data.frame(cbind(cells, group))
fwrite(cellgroups, file='cellgroups.csv')
```


\pagebreak

## Uncommon, but not rare, genes

```{r,fig.width=6,fig.height=8}
print(incidence.sample <- sample(genes.incidence,12))

incidence.mask <- genes %in% incidence.sample
tmp.dat <- cell.dat[incidence.mask,]
tmp.names <- genes[incidence.mask]
par(mfrow=c(4,3))
for(i in 1:dim(tmp.dat)[1]) {
  tmp <- unlist(tmp.dat[i,])
  hist(tmp,main=paste("Histogram, ",tmp.names[i]))
}
```


```{r,eval=FALSE,fig.width=6,fig.height=8}
distances <- dist(t(incidence.dat), method="euclidean")
clusters <- hclust(distances,method="ward.D")
ggdendrogram(clusters,rotate=TRUE)
```

```{r,eval=FALSE,fig.width=6,fig.height=8}
heatmap(t(incidence.dat))
```

\pagebreak

## Gene Ontology



### Question 1.

Is there are subset of genes that, *a priori* would be preferred to differentiate among cell lineages? For example, perhaps zinc-finger protein expression is discriminatory?

```{r,fig.width=6,fig.height=8}
znf.genes <- genes[grepl('^ZNF',genes)]
znf.dat <- cell.dat[grepl('^ZNF',genes),]
distances <- dist(t(znf.dat), method="euclidean")
clusters <- hclust(distances,method="ward.D")
ggdendrogram(clusters,rotate=TRUE)
```

```{r,fig.width=6,fig.height=8}
print(znf.sample <- sample(znf.genes,12))

znf.mask <- genes %in% znf.sample
tmp.dat <- cell.dat[znf.mask,]
tmp.names <- genes[znf.mask]
par(mfrow=c(4,3))
for(i in 1:dim(tmp.dat)[1]) {
  tmp <- unlist(tmp.dat[i,])
  hist(tmp,main=paste("Histogram, ",tmp.names[i]))
}
```


```{r,fig.width=6,fig.height=8}
heatmap(t(znf.dat))
```


```{r}
kmax <- 20
cluster_znf <- data.frame(nclusters=1:kmax,
                             wss = get_wss(znf.dat,kmax))
breaks <- 1:kmax
ggplot(cluster_znf,aes(x=nclusters,y=wss)) + geom_point() + geom_line() +
  scale_x_continuous(breaks=breaks)
```

There is no good cutoff for ZNF, at least for k up to 20. Let's cut at 7 as biologically relevant

(RBC, WBC=[neutrophils, basophils, eosinophils, macrophages, T-lymphocytes, B-lymphocytes])

```{r}

grouped.znf.dat <- data.frame(t(znf.dat))
names(grouped.znf.dat) <- znf.genes
grouped.znf.dat$Group <- cutree(clusters,k=7)
grouped.znf.dat$Group <- as.factor(grouped.znf.dat$Group)
```

```{r}
colnames.work.dat <-colnames(grouped.znf.dat)
temp <- str_replace_all(colnames.work.dat, '-', '_')
temp.work.dat <- grouped.znf.dat
colnames(temp.work.dat) <- temp

m <- length(temp)-1
tbls <- data.frame(x=1:m)

for(i in 1:m){
  fmla <- as.formula(paste(temp[i], "~ Group"))
  aov.tmp <-aov(fmla,data=temp.work.dat)
  aov.tbl <- summary(aov.tmp)
  tbls$pvalue[i] <- aov.tbl[[1]][1,'Pr(>F)']
  tbls$fmla[i] <- temp[i]
}

# Use R to adjust p values
# c("holm", "hochberg", "hommel", "bonferroni", "BH", "BY",
#   "fdr", "none")
tbls$p.adjust <- p.adjust(tbls$pvalue, method = "fdr")
#bonferroni seems to be more conservative for this case
#tbls$p.adjust <- p.adjust(tbls$pvalue, method = "bonferroni")
# code for visualizing BH FDR, but probably not needed
#F.znf <- sort(tbls$pvalue)
#plot(F.znf,type='l')
sig.pvals.znf <- tbls[which(tbls$p.adjust<=0.05),]
fwrite(sig.pvals.znf, file='genesofinterest,znf.csv')
# write cell/group to excel
cells.znf <- rownames(temp.work.dat)
group.znf <- temp.work.dat$Group
cellgroups.znf <- as.data.frame(cbind(cells.znf, group.znf))
fwrite(cellgroups, file='cellgroups.znf.csv')
```

Starting with 438 genes with the text `ZNF` (for zinc finger protein), and assigning 7 groups based on the number of blood cell, we performed clustering and analysis of variance as described above. Only 38 genes of the 438 achieved an FDR adjusted p-value of less than 0.05. These results are exported as  `cellgroups.znf.csv` and `genesofinterest,znf.csv`